### ISSUE TITLE
'zpool attach' on a log device initiates pool resilver

### ISSUE THREAD

**Issue by DeHackEd** [openzfs/zfs#1633]
<pre># zpool status
  pool: mailstore4
 state: ONLINE
  scan: scrub repaired 0 in 0h46m with 0 errors on Thu Jul 25 22:34:54 2013
config:

    NAME               STATE     READ WRITE CKSUM
    mailstore4         ONLINE       0     0     0
      mirror-0         ONLINE       0     0     0
        Slot-0         ONLINE       0     0     0
        Slot-1         ONLINE       0     0     0
      mirror-1         ONLINE       0     0     0
        Slot-2         ONLINE       0     0     0
        Slot-3         ONLINE       0     0     0
      mirror-2         ONLINE       0     0     0
        Slot-4         ONLINE       0     0     0
        Slot-5         ONLINE       0     0     0
      mirror-3         ONLINE       0     0     0
        Slot-6         ONLINE       0     0     0
        Slot-7         ONLINE       0     0     0
      mirror-4         ONLINE       0     0     0
        Slot-8         ONLINE       0     0     0
        Slot-9         ONLINE       0     0     0
    logs
      SSD-Left-part5   ONLINE       0     0     0
    cache
      SSD-Left-part3   ONLINE       0     0     0
      SSD-Right-part3  ONLINE       0     0     0

# zpool attach mailstore4 SSD-Left-part5 SSD-Right-part5
# zpool status
  pool: mailstore4
 state: ONLINE
status: One or more devices is currently being resilvered.  The pool will
    continue to function, possibly in a degraded state.
action: Wait for the resilver to complete.
  scan: resilver in progress since Wed Aug  7 11:57:37 2013
    1.09G scanned out of 578G at 278M/s, 0h35m to go
    0 resilvered, 0.19% done
config:

    NAME                 STATE     READ WRITE CKSUM
    mailstore4           ONLINE       0     0     0
      mirror-0           ONLINE       0     0     0
        Slot-0           ONLINE       0     0     0
        Slot-1           ONLINE       0     0     0
      mirror-1           ONLINE       0     0     0
        Slot-2           ONLINE       0     0     0
        Slot-3           ONLINE       0     0     0
      mirror-2           ONLINE       0     0     0
        Slot-4           ONLINE       0     0     0
        Slot-5           ONLINE       0     0     0
      mirror-3           ONLINE       0     0     0
        Slot-6           ONLINE       0     0     0
        Slot-7           ONLINE       0     0     0
      mirror-4           ONLINE       0     0     0
        Slot-8           ONLINE       0     0     0
        Slot-9           ONLINE       0     0     0
    logs
      mirror-5           ONLINE       0     0     0
        SSD-Left-part5   ONLINE       0     0     0
        SSD-Right-part5  ONLINE       0     0     0
    cache
      SSD-Left-part3     ONLINE       0     0     0
      SSD-Right-part3    ONLINE       0     0     0

errors: No known data errors
</pre>

Note that attaching the second disk to the log initiated a full pool scrub. After it was finished:

<pre>
# zpool status
  pool: mailstore4
 state: ONLINE
  scan: resilvered 0 in 0h26m with 0 errors on Wed Aug  7 12:24:11 2013
config:
    ...
</pre>

So that was a pretty thorough (yet ultimately superfluous) scan it did.

**Comment by behlendorf** [comment]
@DeHackEd do you know if this is still an issue?

**Comment by DeHackEd** [comment]
Tested with ba779f7 and it's still doing it.
```
# zpool add dehacked log /dev/vol00/zfstest2
# zpool attach dehacked  /dev/vol00/zfstest2  /dev/vol00/zfstest3
# zpool status

  pool: dehacked
 state: ONLINE
status: One or more devices is currently being resilvered.  The pool will
	continue to function, possibly in a degraded state.
action: Wait for the resilver to complete.
  scan: resilver in progress since Sun Feb 11 11:23:31 2018
	307G scanned at 1.86G/s, 307G issued at 1.86G/s, 7.50T total
	0B resilvered, 3.99% done, 0 days 01:06:10 to go
config:

	NAME                                           STATE     READ WRITE CKSUM
	dehacked                                       ONLINE       0     0     0
	  raidz2-0                                     ONLINE       0     0     0
	    wwn-0x50014ee26471e85d                     ONLINE       0     0     0
	    scsi-SATA_WDC_WD40EFRX-68_WD-WCC7K5DSJ01R  ONLINE       0     0     0
	    wwn-0x50014ee20ec0109d                     ONLINE       0     0     0
	    wwn-0x50014ee2b96a83b0                     ONLINE       0     0     0
	    wwn-0x5000c500a4d0debe                     ONLINE       0     0     0
	    scsi-SATA_ST4000VN008-2DR_ZDH1NDXL         ONLINE       0     0     0
	    wwn-0x5000c500a4d0f1ba                     ONLINE       0     0     0
	    wwn-0x5000c500a4d15d50                     ONLINE       0     0     0
	logs
	  mirror-1                                     ONLINE       0     0     0
	    zfstest2                                   ONLINE       0     0     0
	    zfstest3                                   ONLINE       0     0     0

errors: No known data errors
```

**Comment by scineram** [comment]
I think this the same issue @avg-I mentioned in the September 2019 meeting: a log device resilver scans the entire pool instead of just the ZIL blocks.

**Comment by avg-I** [comment]
I think that it would be nice to optimize the use-case.
Unfortunately, I am out of time to work on this now.

**Comment by tcely** [comment]
Is removing the (degraded) log mirror from the pool, then adding a new mirror as a log device a valid workaround for this bug?

**Comment by amotin** [comment]
This is not a bug.  Resilver on attach is started to recreate a redundancy of any existing ZIL metadata.  It might be needed if we have some datasets which were not mounted since import and which ZILs are not replayed.  The only thing I think might be added there is a check for any datasets in this state to skip the resilver in some cases, but I am not sure it would not be a layering violation, messing up the code.

@tcely Removal of SLOG drops any ZIL it stores, which might result in an undesired data loss if there is some unreplayed ZIL.

**Comment by tcely** [comment]
Dropping the ZIL without ensuring that it has been replayed would be a bug, in my opinion.

Also, your argument that there is a need to scan the entire pool doesn't make any sense. All of the data needed to recreate the redundancy of the mirror set exists within the degraded mirror set, it should not involve reading already stored data because the log only exists to ensure that the data is stored correctly in the first place.

Any ZIL data that hasn't been replayed, without caring about the mounted state of the dataset, still exists inside the ZIL itself, not in some other part of the pool.

**Comment by amotin** [comment]
> Dropping the ZIL without ensuring that it has been replayed would be a bug, in my opinion.

True, but that is what it is since the day one.  ZILs are replayed during datasets mount.  For encrypted datasets ZIL can not be replayed before that even in theory.  You are welcome to implement ZIL blocks relocation in case of device removal, similar to general case device removal that was added much later.

> Also, your argument that there is a need to scan the entire pool doesn't make any sense. All of the data needed to recreate the redundancy of the mirror set exists within the degraded mirror set, it should not involve reading already stored data because the log only exists to ensure that the data is stored correctly in the first place.

That is how resilver works in ZFS.  It scans all the pool metadata to find all the block pointers with their checksums, but resilvers only the device specified.  I believe "initiated a full pool scrub" in the original report is incorrect.

**Comment by avg-I** [comment]
I do not follow the rationale for closing this issue.
SLOG is being converted from a single vdev to a mirror or from an N-way to N+1 mirror and we need to check every block in the pool for that?

Even if there are still log blocks on non-log vdevs, do we really have to check every block in the pool?
I mean:
1. attaching a log device means only that the log mirror must become valid, not that all log chains must be scrubbed;
2. even if we really want to scrub all the log chains, we could do just that, without traversing other blocks.

More generally, the issue described here is definitely not a bug, but it's a very sub-optimal behavior.
Just my opinion.

**Comment by amotin** [comment]
> do we really have to check every block in the pool?

@avg-I Do you have evidences that we really check every block on the pool?  As part of **resilver** we should scan pool's metadata looking for block pointers, but that should be all.  We should not scan data on other devices.

**Comment by avg-I** [comment]
I probably incorrectly worded my comment, but I think that we visit a lot more block pointers beyond what is strictly required for ZIL.

**Comment by amotin** [comment]
Yes, we likely are.  But in this case more should be better than less.  I'll reopen it from this perspective, if you prefer.

**Comment by tcely** [comment]
> But in this case more should be better than less.

Having just experienced this myself, more is not better. It was, in fact, much worse.

My entire pool spent 5.4 days in a degraded state to resilver a 24 GiB log mirror set. In the end, 0B were actually resilvered.

This could have been completed in seconds instead because according to my `zpool iostat -v` output for that pool, the degraded log mirror spent most of that time with about 8 MiB in use.
